pyramid_tm/__init__.py

```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+            # try to set it as soon as possible so that it's available
+            # in the request factory and elsewhere if people want it
+            # note: set all of these values here as they are cleared after
+            # each attempt
+            environ = request.environ
+            environ['tm.active'] = True
+            environ['tm.attempt'] = number
+            environ['tm.attempts'] = attempts
+
+            # do not touch request.tm until we've set it active or it'll raise
+            if manager is None:
+                manager = request.tm
+
+            # if we are not on the first attempt then we should start
+            # with a new request object and throw away any changes to
+            # the old object, however we do this carefully to try and
+            # avoid extra copies of the body
+            if number > 0:
+                # try to make sure this code stays in sync with pyramid's
+                # router which normally creates requests
+                request = request_factory(environ)
+                request.tm = manager
+                request.registry = registry
+                request.invoke_subrequest = orig_request.invoke_subrequest
+                apply_request_extensions(request, extensions=request_extensions)
+
+            # push the new request onto the threadlocal stack
+            # this should be safe unless someone is doing something
+            # really funky
+            request_manager.push({
+                'request': request,
+                'registry': registry,
+            })
+
             try:
-                manager.begin()
-                # make_body_seekable will copy wsgi.input if necessary,
-                # otherwise it will rewind the copy to position zero
-                if attempts != 1:
-                    request.make_body_seekable()
-                t = manager.get()
-                if userid:
-                    userid = native_(userid, 'utf-8')
-                    t.setUser(userid, '')
+                t = manager.begin()
+
+                # do not address the authentication policy until we are within
+                # the transaction boundaries
+                if annotate_user:
+                    userid = request.unauthenticated_userid
+                    if userid:
+                        userid = native_(userid, 'utf-8')
+                        t.setUser(userid, '')
                 try:
                     t.note(native_(request.path_info, 'utf-8'))
                 except UnicodeDecodeError:
                     t.note("Unable to decode path as unicode")
+
                 response = handler(request)
                 if manager.isDoomed():
-                    raise AbortResponse(response)
+                    raise AbortWithResponse(response)
+
+                # check for a squashed exception and handle it
+                # this would happen if an exception view was invoked and
+                # rendered an error response
```

Comment:
Only if the logic in their exception view can't be rolled back. If that's the case then they can use the predicates as well as `pyramid_tm.exc_is_retryable`, or check the environ values to see if it's the last attempt. Otherwise exception views can just execute like normal.